### PR TITLE
Add url query param (has_permission_check=true) to check if a user has access to call an API

### PR DESCRIPTION
**Pull request by pranathz**
### Description

This PR Adds support to simulate permission checks on existing REST APIs using a query parameter.
* Category (New feature)
* Why these changes are required?
 Previously, there was no way to preview whether a specific action request would be allowed or denied without actually executing it. This change introduces a safe way to check permissions in advance by simply adding a query parameter (simulate_authz=true) to the existing APIs.
* What is the old behavior before changes and new behavior after changes?
   Old Behavior: No support to preview permission results; requests were either executed or blocked.
   New Behavior: When simulate_authz=true is present, the request does not execute but instead returns whether the action would be allowed and lists any missing privileges.

### Issues Resolved
Related to [5468](https://github.com/opensearch-project/security/issues/5468)

**API Design:**
You can add the simulate_authz parameter to any existing REST endpoint.
For e.g.,    `GET /_cluster/health?simulate_authz=true
`

Response format if action is allowed
```
{
  "accessAllowed" : true,
  "missingPrivileges" : [ ]
}
```
Response format if action is not allowed

```
{
  "accessAllowed" : false,
  "missingPrivileges" : ["cluster:monitor/health"]
}
```
### Testing
Integration test included 

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [ ] New Roles/Permissions have a corresponding security dashboards plugin PR
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md)
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/security/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java [hunk 1]
```diff
@@ -388,6 +389,21 @@ private <Request extends ActionRequest, Response extends ActionResponse> void ap
                 log.debug(pres.toString());
             }
 
+            String isSimulation = threadContext.getHeader(ConfigConstants.SECURITY_SIMULATE_AUTHZ_PARAM);
+            if ("true".equals(isSimulation)) {
```
Comment (nagarajg17) [inline src/main/java/org/opensearch/security/filter/SecurityFilter.java]: nit: We can just convert isSimulation to Boolean

File: src/main/java/org/opensearch/security/support/ConfigConstants.java [hunk 2]
```diff
@@ -159,6 +159,11 @@ public class ConfigConstants {
     public static final String SECURITY_AUTHCZ_IMPERSONATION_DN = SECURITY_SETTINGS_PREFIX + "authcz.impersonation_dn";
     public static final String SECURITY_AUTHCZ_REST_IMPERSONATION_USERS = SECURITY_SETTINGS_PREFIX + "authcz.rest_impersonation_user";
 
+    public static final String SECURITY_SIMULATION_API_ENABLED = SECURITY_SETTINGS_PREFIX + "simulation_api.enabled";
+    public static final boolean SECURITY_SIMULATION_API_ENABLED_DEFAULT = true;
```
Comment (nagarajg17) [inline src/main/java/org/opensearch/security/support/ConfigConstants.java]: Is it used anywhere?

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java [hunk 3]
```diff
@@ -300,4 +302,38 @@ public void testAllowlistedApiWithoutExtraSlash() throws Exception {
...
+            nonAdminCredsHeader
+        );
+
+        rh.sendAdminCertificate = false;
+        response = rh.executeGetRequest("_cluster/health?" + ConfigConstants.SECURITY_SIMULATE_AUTHZ_PARAM + "=true", nonAdminCredsHeader);
+        assertThat(response.getBody(), response.getStatusCode(), equalTo(HttpStatus.SC_OK));
+        assertTrue(response.getBody().contains("\"accessAllowed\":"));
+        assertTrue(response.getBody().contains("\"missingPrivileges\":"));
+
+        response = rh.executeGetRequest("_search?" + ConfigConstants.SECURITY_SIMULATE_AUTHZ_PARAM + "=true", nonAdminCredsHeader);
```
Comment (nagarajg17) [inline src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: `_search` is not allowlisted like above for `_cluster/health`?

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java [hunk 4]
```diff
@@ -300,4 +302,38 @@ public void testAllowlistedApiWithoutExtraSlash() throws Exception {
...
+
+    /**
+     * Tests that the simulate_authz param works correctly.
+     * When simulate_authz=true is added to a request, returns
+     * whether the request would be allowed or Denied, without actually executing the request.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testSimulateAuthzParam() throws Exception {
```
Comment (nagarajg17) [inline src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: - We can add one test with value not as true?
- One more test - without this query param this new code flow shouldn't be invoked (to test there is no regression)

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 1 above)
Comment (shikharj05) [inline src/main/java/org/opensearch/security/filter/SecurityFilter.java]: nice! maybe extract all this handling out to a different Function.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 2 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: Not used, will remove in the next commit.

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 1 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: Ack.

Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: Ack.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 2 above)
Comment (cwperks) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: Agree, I don't think feature flag is needed here.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java [hunk 5]
```diff
@@ -159,6 +159,11 @@ public class ConfigConstants {
...
+
+    public static final String SECURITY_SIMULATE_AUTHZ_PARAM = "simulate_authz";
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/support/ConfigConstants.java]: IMO I'd prefer to name this `has_permission` or `can_perform` since it is not a simulation, but rather asking whether the user has permission without actually performing the action.

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 1 above)
Comment (cwperks) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: This looks good for the general case, but there may be a few edge cases that we need to test out or consider as well:

1. I think this will be ignored for superuser because their requests never hit here. We can either choose to exit early [here](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java#L170-L175) or have a log message saying that the param is ignored for superuser
2. For users with DLS/FLS/Field Masking restrictions, should we also consider returning what the restrictions may be if the simulate request would apply the restrictions? i.e. user has DLS restrictions on an index so they are forbidden from writing even if they have an invalid role that says they can write to an index.

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java (hunk 3 above)
Comment (pranathz) [reply src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: correct, I intentionally didn’t add `_search` to the allowlist to ensure it returns a forbidden error. This validates the negative scenario where the action is not allowlisted

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java (hunk 4 above)
Comment (pranathz) [reply src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: Added both tests.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 5 above)
Comment (nagarajg17) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: `has_permission=true` doesn’t imply the intent i.e checking permissions, instead gives meaning like permission is there/true. How about `has_permissions_check` or `has_permissions_validate`? Intent is clear here without ambiguity

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 1 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: Based on SecurityRestFilter this header will only be present if simulate_authz param is present in request. We can simplify the if block to `!Strings.isNullOrEmpty(isSimulation)`.

File: src/main/java/org/opensearch/security/filter/SecurityRestFilter.java [hunk 6]
```diff
@@ -164,6 +164,10 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
                 return;
             }
 
+            boolean isSimulation = request.paramAsBoolean(ConfigConstants.SECURITY_SIMULATE_AUTHZ_PARAM, false);
+            if (isSimulation) {
+                threadContext.putHeader(ConfigConstants.SECURITY_SIMULATE_AUTHZ_PARAM, Boolean.TRUE.toString());
+            }
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/filter/SecurityRestFilter.java:174]: Have we tested the behavior in multi-node scenario?

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 5 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: how about `dry_run`?

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 2 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: this feature flag is not needed. We can remove this

File: src/main/java/org/opensearch/security/filter/SecurityRestFilter.java [hunk 7]
```diff
@@ -164,6 +164,10 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
                 return;
             }
 
+            boolean isSimulation = request.paramAsBoolean(ConfigConstants.SECURITY_SIMULATE_AUTHZ_PARAM, false);
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/filter/SecurityRestFilter.java]: no need to check paramAsBoolean. We can simply consume param and if the param exists, we know its simulation, or a dry_run, and save some network bytes :D

Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityRestFilter.java]: okay, I'll update

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 1 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: > For users with DLS/FLS/Field Masking restrictions, should we also consider returning what the restrictions may be if the simulate request would apply the restrictions? i.e. user has DLS restrictions on an index so they are forbidden from writing even if they have an invalid role that says they can write to an index.

DLS is a read-time control.it does not apply to write operations. So, even if a user has DLS restrictions on an index, they are not forbidden from writing if their role grants write permissions. For your example, even if we wanted to return what the restrictions would be, it wouldn’t make sense for write operations, since DLS has no effect on writes.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 5 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: I agree with Nagaraj’s point — has_permission=true sounds more like stating a fact rather than performing a check. IMO, `has_permissions_check` feels the more clearer and unambiguous, it clearly conveys that we are checking permissions without actually performing the action.

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java [hunk 8]
```diff
@@ -300,4 +303,79 @@ public void testAllowlistedApiWithoutExtraSlash() throws Exception {
...
+     * @throws Exception
+     */
+    @Test
+    public void testHasPermissionCheckParam() throws Exception {
+        setup();
+
+        // ALLOWLIST GET /_cluster/health
+        rh.keystore = "restapi/kirk-keystore.jks";
+        rh.sendAdminCertificate = true;
+        response = rh.executePutRequest(
```
Comment (cwperks) [inline src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: Does this test do anything with the first response?

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java [hunk 9]
```diff
@@ -300,4 +303,79 @@ public void testAllowlistedApiWithoutExtraSlash() throws Exception {
...
+        );
+
+        rh.sendAdminCertificate = false;
+        response = rh.executeGetRequest("_search?" + ConfigConstants.SECURITY_HAS_PERMISSION_CHECK_PARAM + "=false", nonAdminCredsHeader);
+        assertThat(response.getBody(), response.getStatusCode(), equalTo(HttpStatus.SC_OK));
+        assertFalse(response.getBody().contains("\"accessAllowed\":"));
+        assertFalse(response.getBody().contains("\"missingPrivileges\":"));
+
+        response = rh.executeGetRequest("_cat/nodes?" + ConfigConstants.SECURITY_HAS_PERMISSION_CHECK_PARAM + "=true", nonAdminCredsHeader);
+        assertThat(response.getBody(), response.getStatusCode(), equalTo(HttpStatus.SC_FORBIDDEN));
```
Comment (cwperks) [inline src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: Why is Forbidden expected here instead of the response saying allowed is false with the list of missing permissions?

File: src/main/java/org/opensearch/security/action/simulate/PermissionCheckResponse.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public PermissionCheckResponse(final StreamInput in) throws IOException {
+        this.accessAllowed = in.readBoolean();
+        this.missingPrivileges = in.readSet(StreamInput::readString);
+    }
+
+    @Override
+    public RestStatus status() {
+        return accessAllowed ? RestStatus.OK : RestStatus.FORBIDDEN;
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/action/simulate/PermissionCheckResponse.java]: Status should always be OK since the request was successful. The response will indicate whether action is allowed.

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java [hunk 11]
```diff
@@ -507,6 +511,31 @@ private void attachSourceFieldContext(ActionRequest request) {
...
+        String isSimulation = threadContext.getHeader(ConfigConstants.SECURITY_HAS_PERMISSION_CHECK_PARAM);
+        if (Boolean.parseBoolean(isSimulation)) {
+
+            try {
+
+                @SuppressWarnings("unchecked")
+                Response response = (Response) new PermissionCheckResponse(pres.isAllowed(), pres.getMissingPrivileges());
+                listener.onResponse(response);
+
+                log.info("accessAllowed {}", pres.isAllowed());
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/filter/SecurityFilter.java]: these logs should be changed to debug level and should contain information about what action was simulation run on.

File: src/main/java/org/opensearch/security/filter/SecurityRestFilter.java [hunk 12]
```diff
@@ -164,12 +167,21 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
...
+                threadContext.putHeader(ConfigConstants.SECURITY_HAS_PERMISSION_CHECK_PARAM, Boolean.TRUE.toString());
+            }
             // Authorize Request
             final User user = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
             String intiatingUser = threadContext.getTransient(OPENDISTRO_SECURITY_INITIATING_USER);
             if (userIsSuperAdmin(user, adminDNs)) {
                 // Super admins are always authorized
                 auditLog.logSucceededLogin(user.getName(), true, intiatingUser, requestChannel);
+                if (hasPermissionCheck) {
+                    log.info("Permission check skipped: Super admin has full access");
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/filter/SecurityRestFilter.java]: log level should be changed to debug.

File: src/main/java/org/opensearch/security/support/ConfigConstants.java [hunk 13]
```diff
@@ -159,6 +159,8 @@ public class ConfigConstants {
     public static final String SECURITY_AUTHCZ_IMPERSONATION_DN = SECURITY_SETTINGS_PREFIX + "authcz.impersonation_dn";
     public static final String SECURITY_AUTHCZ_REST_IMPERSONATION_USERS = SECURITY_SETTINGS_PREFIX + "authcz.rest_impersonation_user";
 
+    public static final String SECURITY_HAS_PERMISSION_CHECK_PARAM = "has_permission_check";
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/support/ConfigConstants.java:162]: this is not a configurable constant. it should be moved to the handler class

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java [hunk 14]
```diff
@@ -507,6 +511,31 @@ private void attachSourceFieldContext(ActionRequest request) {
...
+                log.info("missing privileges {}", pres.getMissingPrivileges());
+
+            } catch (Exception e) {
+                log.error("Error creating Permission Check response", e);
+                listener.onFailure(new OpenSearchException("Error creating Permission Check response", e));
```
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/filter/SecurityFilter.java]: i don't think catch block is needed as we are not expecting any checked exceptions here.

File: src/main/java/org/opensearch/security/action/simulate/PermissionCheckResponse.java (hunk 10 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/action/simulate/PermissionCheckResponse.java]: Got it, I'll update

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 11 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: Okay, will change it to debug level

File: src/main/java/org/opensearch/security/filter/SecurityRestFilter.java (hunk 12 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityRestFilter.java]: Sure, will update it to debug level

File: src/main/java/org/opensearch/security/support/ConfigConstants.java (hunk 13 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java:162]: Okay, will move it to the handler class

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java (hunk 8 above)
Comment (pranathz) [reply src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: No, it doesn’t use the first response. I’ll update the test

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 14 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: We still need the catch block because a `ClassCastException` (an unchecked exception) can occur when casting to Response

Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: we only reach inside if block if its a simulation request and if its a simulation request then the expected response is PermissionsCheckResponse and so we would never trigger this for that response.

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java [hunk 15]
```diff
@@ -300,4 +303,67 @@ public void testAllowlistedApiWithoutExtraSlash() throws Exception {
...
+
+        rh.keystore = "restapi/kirk-keystore.jks";
+        rh.sendAdminCertificate = true;
+        response = rh.executeGetRequest(
+            "_cluster/health?has_permission_check=true",
+            nonAdminCredsHeader
+        );
+        rh.sendAdminCertificate = false;
+        assertThat(response.getBody(), response.getStatusCode(), equalTo(HttpStatus.SC_OK));
+        assertTrue(response.getBody().contains("\"accessAllowed\":"));
```
Comment (DarshitChanpura) [inline src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: can you check the actual value as well to see if contains true or false, try a different endpoint where it is explicitly false and contains missingPrivileges.

File: src/main/java/org/opensearch/security/filter/SecurityRestFilter.java (hunk 6 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/filter/SecurityRestFilter.java:174]: Tests use DEFAULT configuration which 1 m, 2d nodes. It satisfies multi-node scenario. resolving comment

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java [hunk 16]
```diff
@@ -300,4 +303,67 @@ public void testAllowlistedApiWithoutExtraSlash() throws Exception {
...
+        assertTrue(response.getBody().contains("\"missingPrivileges\":"));
+
+
+        rh.sendAdminCertificate = true;
```
Comment (cwperks) [inline src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java:352]: Is the following block related to the feature? IMO I don't think its necessary to call a separate API and assert that it doesn't contain the `accessAllowed` or `missingPrivileges` in the body. There are many other tests that cover general behavior.

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 14 above)
Comment (pranathz) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: got it, will change

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java (hunk 16 above)
Comment (pranathz) [reply src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java:352]: Added to verify has_permission_check=false triggers normal flow and doesn’t affect existing behavior.

File: src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java (hunk 15 above)
Comment (pranathz) [reply src/test/java/org/opensearch/security/filter/SecurityRestFilterTests.java]: Screenshot 2025-07-25 at 12 01 14 PM

**Review by nagarajg17 — APPROVED** [review]

**Review by cwperks — DISMISSED** [review]

**Review by cwperks — APPROVED** [review]
